pkg/tfbridge/schema.go

```diff
@@ -1032,83 +1028,115 @@ func propagateDefaultAnnotations(oldInput, newInput resource.PropertyValue, tfs
+	case oldInput.IsString() && newState.IsString():
+		// If this value has a StateFunc, its state value may not be compatible with its
+		// input value. Ignore the difference.
 		if tfs != nil && tfs.StateFunc() != nil {
-			oldStr = tfs.StateFunc()(oldStr)
+			return oldInput, tfs.StateFunc()(oldInput.StringValue()) == newState.StringValue()
 		}
-		return oldStr == newStr
+		return newState, oldInput.StringValue() == newState.StringValue()
 	default:
-		return oldInput.DeepEquals(newInput)
+		return newState, oldInput.DeepEquals(newState)
 	}
 }
 
-func extractInputsFromOutputs(oldInputs, outs resource.PropertyMap,
-	tfs shim.SchemaMap, ps map[string]*SchemaInfo, isRefresh bool) (resource.PropertyMap, error) {
+func extractSchemaInputs(state resource.PropertyValue, tfs shim.Schema, ps *SchemaInfo,
+	rawNames bool) resource.PropertyValue {
 
-	inputs := make(resource.PropertyMap)
-	for name, value := range outs {
-		// If this property is not an input, ignore it.
-		_, sch, _ := getInfoFromPulumiName(name, tfs, ps, false)
-		if sch == nil || (!sch.Optional() && !sch.Required()) {
-			continue
+	contract.Assert(tfs == nil || !tfs.Computed())
+
+	switch {
+	case state.IsArray():
+		etfs, eps := elemSchemas(tfs, ps)
+
+		inputs := make([]resource.PropertyValue, len(state.ArrayValue()))
+		for i, e := range state.ArrayValue() {
+			inputs[i] = extractSchemaInputs(e, etfs, eps, rawNames)
+		}
+		return resource.NewArrayProperty(inputs)
+	case state.IsObject():
+		var tfflds shim.SchemaMap
+		if tfs != nil {
+			if res, isres := tfs.Elem().(shim.Resource); isres {
+				tfflds = res.Schema()
+			}
+		}
+		var psflds map[string]*SchemaInfo
+		if ps != nil {
+			psflds = ps.Fields
 		}
 
-		// Otherwise, copy it to the result.
-		copy, err := copystructure.Copy(value)
-		if err != nil {
-			return nil, err
+		inputs := resource.PropertyMap{}
+		for pulumiName, value := range state.ObjectValue() {
+			_, etfs, eps := getInfoFromPulumiName(pulumiName, tfflds, psflds, rawNames || useRawNames(tfs))
+			if etfs != nil && !etfs.Computed() && (etfs.Required() || etfs.Optional()) {
+				inputs[pulumiName] = extractSchemaInputs(value, etfs, eps, rawNames || useRawNames(tfs))
+			}
 		}
-		inputs[name] = copy.(resource.PropertyValue)
+
+		return resource.NewObjectProperty(inputs)
+	default:
+		return state
 	}
+}
+
+func extractInputsFromOutputs(oldInputs, outs resource.PropertyMap,
+	tfs shim.SchemaMap, ps map[string]*SchemaInfo, isRefresh bool) (resource.PropertyMap, error) {
 
-	// Propagate default annotations from the old inputs.
 	sch := (&schema.Schema{
 		Elem: (&schema.Resource{
 			Schema: tfs,
 		}).Shim(),
 	}).Shim()
 	pss := &SchemaInfo{Fields: ps}
-	propagateDefaultAnnotations(
-		resource.NewObjectProperty(oldInputs), resource.NewObjectProperty(inputs), sch, pss, !isRefresh)
 
-	return inputs, nil
+	var inputs resource.PropertyValue
+	if isRefresh {
+		// If this is a refresh, only extract new values for inputs that are already present.
```

Comment:
>  I think perhaps that case is taken care of by "guaranteeing" that these will be present in inputs even though the user didn't provide them. 

This is correct: if a property is populated using a default value (whether that value is sourced from the Pulumi overlays or TF schema), it will be present in the inputs returned from `Check`, and its name will be present in the `__defaults` list of its containing object.

> Slightly more esoterically, if a new optional input property is added to a resource, and its value is set to something non-default outside of Pulumi - will that be handled correctly here? Am I right that it won't get refreshed immediately, but if an update is performed after the refresh, a default will get populated (will that try to overwrite the value of the resource) and then a second refresh will pick up the value?

I think it's helpful to more crisply define what this scenario, the TF behavior, and the Pulumi behavior with the proposed changes. I'll break this down into two cases: the case of an optional pure input and the case of an optional inout. In both cases, I think that the behavior is a little bit different than what you've described.

In the first case, both TF and Pulumi will detect a difference between the actual and desired state of the resource during the next apply/update. A value will be populated for the new property, and each tool will note the addition of the property. Because the value is a pure input, the external non-default value ("its value is set to something non-default outside of Pulumi") will not be taken into account be either told.

The second case is similar, but the detected different will be between the new input and the external non-default value. Even though a `pulumi refresh` will not pull a new value in for the input, it will pull a new value in for the state, and the bridge's implementation of `Diff` will detect this difference just as TF would. 

> Should all pure inputs always be copied over, even if they weren't in the old state? (This making this a superset of the read case, instead of just an unrelated set vs read)?

I don't think so. That would run into this problem from the PR description:

> certain combinations of output property values may be allowed where
the same combinations of input property values are not

The current approach allows us to be more confident that refresh will produce a legal combination of inputs.